### DISCUSSION TITLE
Near Sign transaction

### DISCUSSION THREAD

**Discussion by headria** [trustwallet/wallet-core#1893] [Q&A]
Hello. I want to sign Near transaction and broadcast it to rpc near. but all of the time I've got an error from rpc server. I will share the fun signer that I have written on Kotlin.

` 

        private fun signTransactionNEAR() {

        val transferAction = NEAR.Transfer.newBuilder().apply {
            deposit = ByteString.copyFrom((BigNumber(getExtAmount()) * BigNumber(1E24)).toPlainString().toHexByteArray())
        }.build()

        val signingInput = NEAR.SigningInput.newBuilder().apply {
            this.
            signerId = coinKey.address
            nonce = nearViewAccessKey.nonce
            receiverId = getExtRecipientAddress()

            addActionsBuilder().apply {
                transfer = transferAction
            }

            blockHash = ByteString.copyFrom(Base58.decodeNoCheck(nearBlockDetail.hash))

           //  ByteString.copyFrom(wallet.getKeyForCoin(coin.toTrustWalletCoin).data())
            privateKey = coinKey.privateKey
        }.build()

        val output = AnySigner.sign(signingInput, CoinType.NEAR, Near.parser())
        val signedTransaction: String = Base64.getEncoder().encodeToString(output.signedTransaction.toByteArray())
         nearViewModel.sendRawTransaction(signedTransaction)
    }
`

**Comment by optout21** [comment]
Have you double checked with wallet-core tests?
Seems you are setting SignerId to address, this seems incorrect.

https://github.com/trustwallet/wallet-core/blob/master/android/app/src/androidTest/java/com/trustwallet/core/app/blockchains/near/TestNEARSigner.kt
https://github.com/trustwallet/wallet-core/blob/master/tests/NEAR/SignerTests.cpp

**Reply by headria** [reply]
The SignerId is ok. because Near blockchain is supports this kind of address.

**Reply by headria** [reply]
@catenocrypt ??

**Reply by headria** [reply]
the problem was with the deposit when I multiply it by 30 ( * 10^30) the transaction was ok. but Its have another problem with the deposit. 
after signing the transaction, the deposit change to an integer number. (ex. I want to send "1000000000000000000000000" which is equal to 1 NEAR, in the final data after signing it changed to 1, which is incorrect and it should be to 1000000000000000000000000 ).

**Reply by headria** [reply]
I will share the result of the correct transaction. and our postman result.

**Reply by headria** [reply]
Website near transaction result for send 5 NEAR,

` 
     "transaction": {
      "signer_id": "arab2.testnet",
      "public_key": "ed25519:DK2DSND8dtEVCxM4miSuKPCNmz7gm6CfWwq6DXfuGrte",
      "nonce": 78650022000002,
      "receiver_id": "b6ea122313da14bb542a3622f36dda184a1b7c9bb1550bb022623e3ee76f736f",
      "actions": [
        {
          "Transfer": {
            "deposit": "5000000000000000000000000"
          }
        }
      ],
      "signature": "ed25519:2AtP1orhrYcTztWRq4tLTx8U7nsQgaW6qi8vkbKRTcxWvLcWN3Zb9mnuPAA3fJKUx3E18GiiZcXM4JSSdWXkYTpp",
      "hash": "5c1xNKLTG4stcAvzYAYwa4E9Gdio6RSqJXx7vMH7rL5s"
    },
`


our transaction result for send 1 NEAR 

`
  "transaction": {
            "signer_id": "1e350eb8eb9c24deec11c12affca191cfa012d76ec0c6eed6c0e0bd03e3941ee",
            "public_key": "ed25519:32vA1Jf5LodSphnxDYpBiLWLmgykA88nEg3psPLhzLBw",
            "nonce": 5,
            "receiver_id": "b6ea122313da14bb542a3622f36dda184a1b7c9bb1550bb022623e3ee76f736f",
            "actions": [
                {
                    "Transfer": {
                        "deposit": "16"
                    }
                }
            ],
            "signature": "ed25519:7enQ2oHprG1XUeznmmhqjyHtgdgk6cuQvzjpbsTCUgAriyaAzwFpXivCK2GA8zfHuH4Pe8pw5p7mMYPfK9Ju1s8",
            "hash": "GGuTR6ks6b6FAjxcbK5XkECKrzzTW1Ss9W5HZnXZw6Fr"
        }
`


please check the deposit field.

**Reply by lamnd91** [reply]
deposit value should be reversed with fixed length = 16. Here is the code in my project, you should try it later
```
fun BigDecimal.toNearByteString() = run {  
        //Convert to Yocto near  
        val amountBigInt = scaleByPowerOfTen(24).toBigInteger()  
        val amountByteArray = amountBigInt.toByteArray()  
        val bytes = amountByteArray.reversedArray().copyOf(16)  
        ByteString.copyFrom(bytes)  
    }
```
